### PR TITLE
Adds an Instrumentation Span Handler to Langfuse

### PR DESCRIPTION
**Pull request by jonathanhliu21**
# Description

Adds `LangfuseSpanHandler`, allowing people to use Langfuse observability with LlamaIndex's instrumentation features instead of using the legacy callback API.

Screenshot 2024-08-16 at 9 18 08 PM
Screenshot 2024-08-16 at 9 18 30 PM

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

Not a new package.

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
  The tests generate warnings because the Langfuse python library uses deprecated Pydantic methods.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

**Review by nerdai — COMMENTED** [review]
Thanks @jonathanhliu21! Excited to see `Langfuse` integration into our instrumentation module.

I notice that we're writing and reading to `_span_map` attribute without using any locking mechanics. Curious to know if you've tested this integration for co-routine safety by checking that traces/spans are emitted as expected when invoking async methods (i think your example images only show sync methods?)

**Comment by jonathanhliu21** [comment]
> I notice that we're writing and reading to `_span_map` attribute without using any locking mechanics. Curious to know if you've tested this integration for co-routine safety by checking that traces/spans are emitted as expected when invoking async methods (i think your example images only show sync methods?)

I have not tested that yet. I didn't realize that the `_span_map` was not thread safe, so I will add locks.

File: llama-index-integrations/callbacks/llama-index-callbacks-langfuse/pyproject.toml
Comment (nerdai) [inline llama-index-integrations/callbacks/llama-index-callbacks-langfuse/pyproject.toml:1]: can we also add `LangfuseSpanHandler` in the `tool.llamahub.class_authors` section?

**Review by nerdai — APPROVED** [review]
this is great, @jonathanhliu21 -- just one minor comment left re: adding class authors in pyproject.toml

File: llama-index-integrations/callbacks/llama-index-callbacks-langfuse/pyproject.toml
Comment (nerdai) [reply llama-index-integrations/callbacks/llama-index-callbacks-langfuse/pyproject.toml:1]: took care of this

**Pull request merged**